sdk/nodejs/runtime/closure/createClosure.ts

```diff
@@ -748,23 +751,74 @@ function getOrCreateEntry(
         }
-        else if (obj && obj.doNotCapture) {
+
+        if (obj && obj.doNotCapture) {
             // object has set itself as something that should not be captured.
-            entry.json = undefined;
+            return true;
         }
-        else if (isDerivedNoCaptureConstructor(obj)) {
+
+        if (isDerivedNoCaptureConstructor(obj)) {
             // this was a constructor that derived from something that should not be captured.
+            return true;
+        }
+
+        return false;
+    }
+
+    function dispatchAny(): void {
+        if (doNotCapture()) {
+            // We do not want to capture this object.  Explicit set .json to undefined so
+            // that we will see that the property is set and we will simply roundtrip this
+            // as the 'undefined value.
             entry.json = undefined;
+            return;
         }
-        else if (obj === undefined || obj === null ||
+
+        if (obj === undefined || obj === null ||
             typeof obj === "boolean" || typeof obj === "number" || typeof obj === "string") {
             // Serialize primitives as-is.
             entry.json = obj;
+            return;
+        }
+
+        const moduleName = findModuleName(obj);
+        if (moduleName && moduleName[0] !== ".") {
+            // @pulumi modules can't ever be referenced on the 'inside'.  Not only are they not
+            // written to support being used on the inside, but serialization also explicitly
+            // removes these modules from node_modules to trim down the total upload size.
+            //
+            // If the user's code actually explicitly references this, then emit a hard error
+            // to let them know what's going on.
+            if (moduleName.startsWith("@pulumi")) {
+                // Find the 'func' to report this issue against. Note: we will always find some
+                // 'func' in the context chain as the only reason we ever even get into closure
+                // creation is because we're serializing out some function.
+                let func: Function = <any>undefined;
+                for (let i = context.frames.length - 1; i >= 0; i--) {
+                    const frame = context.frames[i];
+                    if (frame.functionLocation) {
+                        func = frame.functionLocation.func;
+                    }
+                }
+
+                throwSerializationError(func, context,
+                    "'@pulumi' modules cannot be referenced inside a function that will execute in the Cloud.");
+            }
+
+            // This name bound to a module, and the module did not opt into having itself captured
+            // by value. In this case serialize it out as a direct 'require' call
+            //
+            // Also, importantly, if this is a reference to a local module (i.e. starts with '.'),
+            // then always capture it as a value.  The reason for this is that otherwise we won't
```

Comment:
Is it correct to treat local modules differently from `@pulumi` modules?  What's the rationale? 

* Local: capture by value
* `@pulumi`: reject

Should these both reject, or both capture by value?  The prose here doesn't make clear (at least to me) why these are different.

Reply:
> Should these both reject, or both capture by value? The prose here doesn't make clear (at least to me) why these are different.

I'll improve the prose.  However, basically, both of these are ones we cannot capture by reference.  We could just default to capturing both by value.  

**However**: it's known ahead of time that capturing @pulumi by value simply will not work (simply because it itself references packages that we won't be uploading).  We could just allow the capturing to seem to succeed (or potentially fail), but the user will most likely then get a cryptic error later on down the line.

The point of erroring early is simply so we can provide good failure path.  In other words, using @pulumi is just going ot fail at this part of our product cycle.  We can either make it fail very cryptically (i.e. errors at runtime, or with suggestions to change how it is imported), or we can detect up front this bad pattern and give a *very good* message about what is going wrong.  

Normally, i'm really not a fan of special casing, or having magic strings.  And i definitely be up for some sort of cleaner mechanism to address this issue.  However, i think this is a very low cost, and *effective* mechanism for quick and clearly helping people avoid something that is almost surely going to otehrwise be a bad experience :)

--

Note: i'm ok with opening a bug to track investigating other things here.  However, i have a real concern that such bugs simply sit around not serving much purpose since they'll almost always (in isolation) be a low priority for us.  I'd rather us just accept this as an acceptable starting point, and we can drive any changes here through new scenarios with clear customers (internal or external) that need those scenarios to work.